syntax = "proto3";
package quite.proto;

import "quite/proto/common.proto";
import "quite/proto/keyboard.proto";
import "quite/proto/mouse.proto";

service ProbeService {
    rpc FindObject(ObjectRequest) returns (ObjectReply) {}
    rpc GetObjectProperty(PropertyRequest) returns (PropertyResponse) {}

    rpc MouseAction(MouseActionRequest) returns (VoidResponse) {}
    rpc CreateScreenshot(ScreenshotRequest) returns (ImageResponse) {};
}

message ObjectRequest {
  ObjectRequest parent = 1;
  string object_name = 2;
}

message ObjectReply {
  uint64 id = 1; // the internal id assigned by the probe
  string type_name = 2;
}

message PropertyRequest {
  uint64 id = 1;
  string property_name = 2;
}

message PropertyResponse {
  string type_name = 1;
  string value = 2;
}

message MouseActionRequest {
  uint64 target_id = 1;
  MouseAction mouse_action = 2;
  MouseButton mouse_button = 3;
  optional KeyboardModifierKey modifier_key = 4;
  optional uint32 delay_ms = 5; // delay between press and release
  optional Vector2F relative_point = 6; // the (click) point relative to the object coordinates. 
}


message VoidResponse {

}

message ImageMetadata {
  uint32 width = 1;
  uint32 height = 2;
}
message ImageResponse {
  ImageMetadata metadata = 1;
  bytes data = 2;
}

message ScreenshotRequest {
  uint64 id = 1;
}
